modules/tensor_mechanics/src/materials/ADViscoplasticityStressUpdate.C

```diff
@@ -0,0 +1,387 @@
+    _effective_inelastic_strain[_qp] += dpsi_dgauge * _dt;
+    // Update creep strain due to currently computed inelastic strain
+    _creep_strain[_qp] += inelastic_strain_increment;
+  }
+
+  computeStressFinalize(inelastic_strain_increment);
+}
+
+template <ComputeStage compute_stage>
+Real
+ADViscoplasticityStressUpdate<compute_stage>::computeReferenceResidual(
+    const ADReal & /*effective_trial_stress*/, const ADReal & gauge_stress)
+{
+  // Use gauge stress for relative tolerance criteria, defined as:
+  // std::abs(residual / gauge_stress) <= _relative_tolerance
+  return MetaPhysicL::raw_value(gauge_stress);
+}
+
+template <ComputeStage compute_stage>
+ADReal
+ADViscoplasticityStressUpdate<compute_stage>::initialGuess(const ADReal & effective_trial_stress)
+{
+  return effective_trial_stress;
+}
+
+template <ComputeStage compute_stage>
+ADReal
+ADViscoplasticityStressUpdate<compute_stage>::maximumPermissibleValue(
+    const ADReal & effective_trial_stress) const
+{
+  return effective_trial_stress * 1.0e6;
+}
+
+template <ComputeStage compute_stage>
+ADReal
+ADViscoplasticityStressUpdate<compute_stage>::minimumPermissibleValue(
+    const ADReal & effective_trial_stress) const
+{
+  return effective_trial_stress;
+}
+
+template <ComputeStage compute_stage>
+Real
+ADViscoplasticityStressUpdate<compute_stage>::computeTimeStepLimit()
+{
+  const Real scalar_inelastic_strain_incr =
+      MetaPhysicL::raw_value(_effective_inelastic_strain[_qp]) -
+      _effective_inelastic_strain_old[_qp];
+
+  if (MooseUtils::absoluteFuzzyEqual(scalar_inelastic_strain_incr, 0.0))
+    return std::numeric_limits<Real>::max();
+
+  return _dt * _max_inelastic_increment / scalar_inelastic_strain_incr;
+}
+
+template <ComputeStage compute_stage>
+void
+ADViscoplasticityStressUpdate<compute_stage>::outputIterationSummary(
+    std::stringstream * iter_output, const unsigned int total_it)
+{
+  if (iter_output)
+  {
+    *iter_output << "At element " << _current_elem->id() << " _qp=" << _qp << " Coordinates "
+                 << _q_point[_qp] << " block=" << _current_elem->subdomain_id() << '\n';
+  }
+  ADSingleVariableReturnMappingSolution<compute_stage>::outputIterationSummary(iter_output,
+                                                                               total_it);
+}
+
+template <ComputeStage compute_stage>
+ADReal
+ADViscoplasticityStressUpdate<compute_stage>::computeResidual(const ADReal & equiv_stress,
+                                                              const ADReal & trial_gauge)
+{
+  const ADReal M = std::abs(_hydro_stress) / trial_gauge;
+  const ADReal dM_dtrial_gauge = -M / trial_gauge;
+
+  const ADReal residual_left = Utility::pow<2>(equiv_stress / trial_gauge);
+  const ADReal dresidual_left_dtrial_gauge = -2.0 * residual_left / trial_gauge;
+
+  ADReal residual = residual_left;
+  _derivative = dresidual_left_dtrial_gauge;
+
+  if (_pore_shape == "spherical")
```

Comment:
Do you want to compare enums with strings in this way?  Sometimes this can be slow, so recently i've been writing Enums like _stabilization in porous_flow/src/actions/PorousFlowActionBase.C

Reply:
Yeah, I didn't think about speed. I was trying to avoid maintaining enums in the header and the source file, but I guess it makes sense for performance.